File: src/modules/shared/domain/zambretti.ts
export enum ZambrettiEnum {
  FALLING_SETTLED_FINE = 1,
  FALLING_FINE_WEATHER = 2,
  FINE_BECOMING_LESS_SETTLED = 3,
  FAIRLY_FINE_SHOWERY_LATER = 4,
  SHOWERY_BECOMING_MORE_UNSETTLED = 5,
  UNSETTLED_RAIN_LATER = 6,
  RAIN_AT_TIMES_WORSE_LATER = 7,
  RAIN_AT_TIMES_BECOMING_VERY_UNSETTLED = 8,
  FALLING_VERY_UNSETTLED_RAIN = 9,
  STEADY_SETTLED_FINE = 10,
  STEADY_FINE_WEATHER = 11,
  FINE_POSSIBLY_SHOWERS = 12,
  FAIRLY_FINE_SHOWERS_LIKELY = 13,
  SHOWERY_BRIGHT_INTERVALS = 14,
  CHANGEABLE_SOME_RAIN = 15,
  UNSETTLED_RAIN_AT_TIMES = 16,
  RAIN_AT_FREQUENT_INTERVALS = 17,
  STEADY_VERY_UNSETTLED_RAIN = 18,
  STEADY_STORMY_MUCH_RAIN = 19,
  RISING_SETTLED_FINE = 20,
  RISING_FINE_WEATHER = 21,
  BECOMING_FINE = 22,
  FAIRLY_FINE_IMPROVING = 23,
  FAIRLY_FINE_POSSIBLY_SHOWERS_EARLY = 24,
  SHOWERY_EARLY_IMPROVING = 25,
  CHANGEABLE_MENDING = 26,
  RATHER_UNSETTLED_CLEARING_LATER = 27,
  UNSETTLED_PROBABLY_IMPROVING = 28,
  UNSETTLED_SHORT_FINE_INTERVALS = 29,
  VERY_UNSETTLED_FINER_AT_TIMES = 30,
  STORMY_POSSIBLY_IMPROVING = 31,
  RISING_STORMY_MUCH_RAIN = 32,
}

enum PressureTrendsEnum {
  FALLING,
  STEADY,
  RISING,
}
const SIGNIFICANT_PRESSURE_CHANGE = 1.6;

export class Zambretti {
  constructor() {}

  private static hasPressureDropped = (current: number, last: number) => {
    return current + SIGNIFICANT_PRESSURE_CHANGE <= last;
  };

  private static hasPressureRaised = (current: number, last: number) => {
    return current >= last + SIGNIFICANT_PRESSURE_CHANGE;
  };

  private static getPressureTrends = (
    currentPressureMeasure: number,
    pastThreeHourPressureMeasure: number = 1015,
  ) => {
    if (
      currentPressureMeasure >= 985 &&
      currentPressureMeasure <= 1050 &&
      Zambretti.hasPressureDropped(
        currentPressureMeasure,
        pastThreeHourPressureMeasure,
      )
    ) {
      return PressureTrendsEnum.FALLING;
    }

    if (
      currentPressureMeasure >= 960 &&
      currentPressureMeasure <= 1033 &&
      !Zambretti.hasPressureDropped(
        currentPressureMeasure,
        pastThreeHourPressureMeasure,
      ) &&
      !Zambretti.hasPressureRaised(
        currentPressureMeasure,
        pastThreeHourPressureMeasure,
      )
    ) {
      return PressureTrendsEnum.STEADY;
    }

    if (
      currentPressureMeasure >= 947 &&
      currentPressureMeasure <= 1030 &&
      Zambretti.hasPressureRaised(
        currentPressureMeasure,
        pastThreeHourPressureMeasure,
      )
    ) {
      return PressureTrendsEnum.RISING;
    }

    throw new Error(
      `Wasn't possible to determine the pressure trend for p=${currentPressureMeasure}`,
    );
  };

  private static calculateZambrettiConstant = (
    currentPressureMeasure: number,
    pressureTrend: PressureTrendsEnum,
  ): number => {
    switch (pressureTrend) {
      case PressureTrendsEnum.FALLING:
        return 127 - 0.12 * currentPressureMeasure;
      case PressureTrendsEnum.STEADY:
        return 144 - 0.13 * currentPressureMeasure;
      case PressureTrendsEnum.RISING:
        return 185 - 0.16 * currentPressureMeasure;
      default:
        throw Error(
          `Wasn't possible to calculate the zambretti value for pressure trend '${pressureTrend}'`,
        );
    }
  };

  static calculateZambretti = (
    currentPressureMeasure: number,
    pastThreeHourPressureMeasure?: number,
  ): ZambrettiEnum => {
    const pressureTrend = this.getPressureTrends(
      currentPressureMeasure,
      pastThreeHourPressureMeasure,
    );

    const zambrettiConstant = Math.round(
      this.calculateZambrettiConstant(currentPressureMeasure, pressureTrend),
    );

    return zambrettiConstant;
  };
}

File: app/lib/core/domain/measure/zambretti.dart
enum ZambrettiEnum {
  fallingSettledFine,
  fallingFineWeather,
  fineBecomingLessSettled,
  fairlyFineShoweryLater,
  showeryBecomingMoreUnsettled,
  unsettledRainLater,
  rainAtTimesWorseLater,
  rainAtTimesBecomingVeryUnsettled,
  fallingVeryUnsettledRain,
  steadySettledFine,
  steadyFineWeather,
  finePossiblyShowers,
  fairlyFineShowersLikely,
  showeryBrightIntervals,
  changeableSomeRain,
  unsettledRainAtTimes,
  rainAtFrequentIntervals,
  steadyVeryUnsettledRain,
  steadyStormyMuchRain,
  risingSettledFine,
  risingFineWeather,
  becomingFine,
  fairlyFineImproving,
  fairlyFinePossiblyShowersEarly,
  showeryEarlyImproving,
  changeableMending,
  ratherUnsettledClearingLater,
  unsettledProbablyImproving,
  unsettledShortFineIntervals,
  veryUnsettledFinerAtTimes,
  stormyPossiblyImproving,
  risingStormyMuchRain,
}

extension ZambrettiExtension on ZambrettiEnum {
  static ZambrettiEnum getZambrettiEnumFromInt(int value) {
    switch (value) {
      case 1:
        return ZambrettiEnum.fallingSettledFine;
      case 2:
        return ZambrettiEnum.fallingFineWeather;
      case 3:
        return ZambrettiEnum.fineBecomingLessSettled;
      case 4:
        return ZambrettiEnum.fairlyFineShoweryLater;
      case 5:
        return ZambrettiEnum.showeryBecomingMoreUnsettled;
      case 6:
        return ZambrettiEnum.unsettledRainLater;
      case 7:
        return ZambrettiEnum.rainAtTimesWorseLater;
      case 8:
        return ZambrettiEnum.rainAtTimesBecomingVeryUnsettled;
      case 9:
        return ZambrettiEnum.fallingVeryUnsettledRain;
      case 10:
        return ZambrettiEnum.steadySettledFine;
      case 11:
        return ZambrettiEnum.steadyFineWeather;
      case 12:
        return ZambrettiEnum.finePossiblyShowers;
      case 13:
        return ZambrettiEnum.fairlyFineShowersLikely;
      case 14:
        return ZambrettiEnum.showeryBrightIntervals;
      case 15:
        return ZambrettiEnum.changeableSomeRain;
      case 16:
        return ZambrettiEnum.unsettledRainAtTimes;
      case 17:
        return ZambrettiEnum.rainAtFrequentIntervals;
      case 18:
        return ZambrettiEnum.steadyVeryUnsettledRain;
      case 19:
        return ZambrettiEnum.steadyStormyMuchRain;
      case 20:
        return ZambrettiEnum.risingSettledFine;
      case 21:
        return ZambrettiEnum.risingFineWeather;
      case 22:
        return ZambrettiEnum.becomingFine;
      case 23:
        return ZambrettiEnum.fairlyFineImproving;
      case 24:
        return ZambrettiEnum.fairlyFinePossiblyShowersEarly;
      case 25:
        return ZambrettiEnum.showeryEarlyImproving;
      case 26:
        return ZambrettiEnum.changeableMending;
      case 27:
        return ZambrettiEnum.ratherUnsettledClearingLater;
      case 28:
        return ZambrettiEnum.unsettledProbablyImproving;
      case 29:
        return ZambrettiEnum.unsettledShortFineIntervals;
      case 30:
        return ZambrettiEnum.veryUnsettledFinerAtTimes;
      case 31:
        return ZambrettiEnum.stormyPossiblyImproving;
      case 32:
        return ZambrettiEnum.risingStormyMuchRain;
      default:
        throw Exception("Invalid integer value: $value");
    }
  }
}
